bika/lims/utils/analysisrequest.py

```diff
@@ -549,3 +469,99 @@ def resolve_rejection_reasons(values):
+    only be sent if the setting in Setup is enabled
+    """
+    sample_id = api.get_id(sample)
+    if not sample.getRejectionReasons():
+        logger.warn("Cannot reject {} w/o rejection reasons".format(sample_id))
+        return
+
+    success, msg = doActionFor(sample, "reject")
+    if not success:
+        logger.warn("Cannot reject the sample {}".format(sample_id))
+        return
+
+    # Generate a pdf with the rejection reasons
+    pdf = get_rejection_pdf(sample)
+
+    # Attach the PDF to the sample
+    filename = "{}-rejected.pdf".format(sample_id)
+    sample.createAttachment(pdf, filename=filename)
+
+    # Do we need to send a notification email?
+    if notify is None:
+        setup = api.get_setup()
+        notify = setup.getNotifyOnSampleRejection()
+
+    if notify:
+        # Compose the email
+        mime_msg = get_rejection_mail(sample)
+        # Attach the pdf
+        mime_msg.attach(to_email_attachment(pdf))
+        # Send the email
+        send_email(mime_msg)
+
+
+def get_rejection_pdf(sample):
+    """Generates a pdf with sample rejection reasons
+    """
+    # Avoid circular dependencies
+    from bika.lims.browser.analysisrequest.reject import \
+        AnalysisRequestRejectPdfView
+
+    # Render the html's rejection document
+    tpl = AnalysisRequestRejectPdfView(sample, api.get_request())
+    html = tpl.template()
+    html = safe_unicode(html).encode("utf-8")
+
+    # Generate the pdf
+    return createPdf(htmlreport=html)
+
+
+def get_rejection_mail(sample):
+    """Generates an email to sample contacts with rejection reasons
+    """
+    # Avoid circular dependencies
+    from bika.lims.browser.analysisrequest.reject import \
+        AnalysisRequestRejectEmailView
+
+    # Render the email body
+    tpl = AnalysisRequestRejectEmailView(sample, api.get_request())
+    email_body = tpl.template()
+
+    def to_valid_email_address(contact):
+        if not contact:
+            return None
+        address = contact.getEmailAddress()
+        if not is_valid_email_address(address):
+            return None
+        name = contact.getFullname() or ""
+        return to_email_address(address, name=name)
+
+    # Get the recipients
+    _to = [sample.getContact()] + sample.getCCContact()
+    _to = map(to_valid_email_address, _to)
+    _to = filter(None, _to)
+
+    # Compose the email
+    lab = api.get_setup().laboratory
+    _from = to_email_address(lab.getEmailAddress(), lab.getName())
+    _subject = _("%s has been rejected") % api.get_id(sample)
+    _subject = to_email_subject(_subject)
+    _body = to_email_body_text(email_body)
+
+    # Create the enclosing message
+    _preamble = "This is a multi-part message in MIME format.\n"
+    mime_msg = MIMEMultipart()
+    mime_msg.preamble = _preamble
+    mime_msg["Subject"] = _subject
+    mime_msg["From"] = _from
+    mime_msg["To"] = ", ".join(_to)
+    mime_msg.attach(_body)
+    return mime_msg
```

Comment:
Would you mind to refactor this to use our API here?
https://github.com/senaite/senaite.core/blob/master/bika/lims/api/mail.py#L175

Reply:
Sorry, this one uses the API already, so maybe just the place below

Reply:
I can, but I need to modify that function, cause although the signature says 
```
:param to_address: List of email or (name, email) pairs
```
https://github.com/senaite/senaite.core/blob/c1f98444f0d33cb5d8f3e12c104c13b4c5ce732a/bika/lims/api/mail.py#L179

And I'd say this is actually not supported: https://github.com/senaite/senaite.core/blob/master/bika/lims/api/mail.py#L187

Reply:
Ah, I see.  Could you fix that please so that it fits for your use-case? I would like that all our email logic relies on that API for further refactorings...

Reply:
done